multidict/_multidict_c.c

```diff
@@ -0,0 +1,854 @@
+{
+    return _multidict_getone(self, key, NULL);
+}
+
+static int
+multidict_mp_as_subscript(_MultiDictObject *self, PyObject *key, PyObject *val)
+{
+    if (val == NULL) {
+        return pair_list_del(self->impl, key);
+    } else {
+        return pair_list_replace(self->impl, key, val);
+    }
+}
+
+static int
+multidict_sq_contains(_MultiDictObject *self, PyObject *key)
+{
+    return pair_list_contains(self->impl, key);
+}
+
+static PyObject *
+mulditict_tp_iter(_MultiDictObject *self)
+{
+    return PyObject_GetIter(multidict_keysview_new((PyObject*)self));
+}
+
+static PyObject *
+multidict_tp_richcompare(PyObject *self, PyObject *other, int op)
+{
+    int cmp = 0;
+
+    if (op != Py_EQ && op != Py_NE) {
+        Py_RETURN_NOTIMPLEMENTED;
+    }
+
+    // TODO: add MultiDictProxy_Check
+    if (MultiDict_Check(other)) {
+        return _multidict_eq(
+            (_MultiDictObject*)self,
+            (_MultiDictObject*)other
+        );
+    }
+
+    cmp = PyObject_IsInstance(other, (PyObject*)&collections_abc_mapping);
+    if (cmp < 0) {
+        return NULL;
+    }
+
+    if (cmp) {
+        if (pair_list_eq_to_mapping(((_MultiDictObject*)self)->impl, other)) {
+            Py_RETURN_TRUE;
+        }
+        Py_RETURN_FALSE;
+    }
+
+    Py_RETURN_NOTIMPLEMENTED;
+}
+
+PyDoc_STRVAR(multidict_getall_doc,
+"Return a list of all values matching the key.");
+
+PyDoc_STRVAR(multidict_getone_doc,
+"Get first value matching the key.");
+
+PyDoc_STRVAR(multidict_get_doc,
+"Get first value matching the key.\n\nThe method is alias for .getone().");
+
+PyDoc_STRVAR(multidict_keys_doc,
+"Return a new view of the dictionary's keys.");
+
+PyDoc_STRVAR(multidict_items_doc,
+"Return a new view of the dictionary's items *(key, value) pairs).");
+
+PyDoc_STRVAR(multidict_values_doc,
+"Return a new view of the dictionary's values.");
+
+/******************** MultiDict ********************/
+
+static int
+multidict_tp_init(_MultiDictObject *self, PyObject *args, PyObject *kwds)
+{
+    self->impl = pair_list_new();
+    if (self->impl == NULL) {
+        return -1;
+    }
+    if (_multidict_extend(self, args, kwds, "MultiDict", 1) < 0) {
+        return -1;
+    }
+    return 0;
+}
+
+static PyObject *
+multidict_add(_MultiDictObject *self, PyObject *args)
+{
+    PyObject *key = NULL,
+             *val = NULL;
+
+    if (!PyArg_UnpackTuple(args, "set", 2, 2, &key, &val)) {
+        return NULL;
+    }
+
+    if (pair_list_add(self->impl, key, val) < 0) {
+        return NULL;
+    }
+
+    Py_RETURN_NONE;
+}
+
+static PyObject *
+multidict_copy(_MultiDictObject *self)
+{
+    _MultiDictObject *new_multidict = NULL;
+
+    PyObject *arg_items = NULL,
+             *items     = NULL;
+
+    new_multidict = PyObject_GC_New(_MultiDictObject, &multidict_type);
+    if (new_multidict == NULL) {
+        return NULL;
+    }
+
+    items = multidict_items(self);
+    if (items == NULL) {
+        goto fail;
+    }
+
+    arg_items = PyTuple_New(1);
```

Comment:
Implementation looks as slow as possible but we can improve it later

Reply:
In TODO